From beeb5456ec957520b52c463efbcfd42228e7a07a Mon Sep 17 00:00:00 2001
From: Uri Mashiach <uri.mashiach@compulab.co.il>
Date: Tue, 19 Sep 2017 10:07:26 +0300
Subject: [PATCH 19/57] eeprom: layout: add field functions read

Read option is missing in the EEPROM layout feature

Add read function to the structures eeprom_layout and eeprom_field.
Add a general binary read function that copy the field content to the
provided buffer without manipulation.

Signed-off-by: Uri Mashiach <uri.mashiach@compulab.co.il>
---
 board/compulab/common/eeprom.c | 107 +++++++++++++++++----------------
 common/eeprom/eeprom_field.c   |  22 +++++++
 common/eeprom/eeprom_layout.c  |  37 ++++++++++++
 include/eeprom_field.h         |   4 ++
 include/eeprom_layout.h        |   2 +
 5 files changed, 119 insertions(+), 53 deletions(-)

diff --git a/board/compulab/common/eeprom.c b/board/compulab/common/eeprom.c
index 1ab81b3541..3e7ae1e31e 100644
--- a/board/compulab/common/eeprom.c
+++ b/board/compulab/common/eeprom.c
@@ -402,16 +402,20 @@ int eeprom_field_update_date(struct eeprom_field *field, uchar *fbuf,
 
 extern struct eeprom_field layout_unknown[1];
 
-#define DEFINE_PRINT_UPDATE(x) eeprom_field_print_##x, eeprom_field_update_##x
+#define DEFINE_FIELD_FUNC(x) eeprom_field_print_##x, eeprom_field_update_##x, \
+			     eeprom_field_read_bin
+
+#define FIELD_FUNC_RES_LAST eeprom_field_print_reserved, \
+			    eeprom_field_update_ascii,   \
+			    eeprom_field_read_bin
 
 #ifdef CONFIG_CM_T3X
 struct eeprom_field layout_legacy[5] = {
-	{ "MAC address",          6, DEFINE_PRINT_UPDATE(mac) },
-	{ "Board Revision",       2, DEFINE_PRINT_UPDATE(bin) },
-	{ "Serial Number",        8, DEFINE_PRINT_UPDATE(bin) },
-	{ "Board Configuration", 64, DEFINE_PRINT_UPDATE(ascii) },
-	{ RESERVED_FIELDS,      176, eeprom_field_print_reserved,
-				     eeprom_field_update_ascii },
+	{ "MAC address",          6, DEFINE_FIELD_FUNC(mac) },
+	{ "Board Revision",       2, DEFINE_FIELD_FUNC(bin) },
+	{ "Serial Number",        8, DEFINE_FIELD_FUNC(bin) },
+	{ "Board Configuration", 64, DEFINE_FIELD_FUNC(ascii) },
+	{ RESERVED_FIELDS,      176, FIELD_FUNC_RES_LAST }
 };
 #else
 #define layout_legacy layout_unknown
@@ -419,61 +423,58 @@ struct eeprom_field layout_legacy[5] = {
 
 #if defined(CONFIG_CM_T3X) || defined(CONFIG_CM_T3517)
 struct eeprom_field layout_v1[12] = {
-	{ "Major Revision",      2, DEFINE_PRINT_UPDATE(bin_ver) },
-	{ "Minor Revision",      2, DEFINE_PRINT_UPDATE(bin_ver) },
-	{ "1st MAC Address",     6, DEFINE_PRINT_UPDATE(mac) },
-	{ "2nd MAC Address",     6, DEFINE_PRINT_UPDATE(mac) },
-	{ "Production Date",     4, DEFINE_PRINT_UPDATE(date) },
-	{ "Serial Number",      12, DEFINE_PRINT_UPDATE(bin_rev) },
-	{ RESERVED_FIELDS,      96, DEFINE_PRINT_UPDATE(reserved) },
-	{ "Product Name",       16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #1", 16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #2", 16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #3", 16, DEFINE_PRINT_UPDATE(ascii) },
-	{ RESERVED_FIELDS,      64, eeprom_field_print_reserved,
-				    eeprom_field_update_ascii },
+	{ "Major Revision",      2, DEFINE_FIELD_FUNC(bin_ver) },
+	{ "Minor Revision",      2, DEFINE_FIELD_FUNC(bin_ver) },
+	{ "1st MAC Address",     6, DEFINE_FIELD_FUNC(mac) },
+	{ "2nd MAC Address",     6, DEFINE_FIELD_FUNC(mac) },
+	{ "Production Date",     4, DEFINE_FIELD_FUNC(date) },
+	{ "Serial Number",      12, DEFINE_FIELD_FUNC(bin_rev) },
+	{ RESERVED_FIELDS,      96, DEFINE_FIELD_FUNC(reserved) },
+	{ "Product Name",       16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #1", 16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #2", 16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #3", 16, DEFINE_FIELD_FUNC(ascii) },
+	{ RESERVED_FIELDS,      64, FIELD_FUNC_RES_LAST }
 };
 #else
 #define layout_v1 layout_unknown
 #endif
 
 struct eeprom_field layout_v2[15] = {
-	{ "Major Revision",            2, DEFINE_PRINT_UPDATE(bin_ver) },
-	{ "Minor Revision",            2, DEFINE_PRINT_UPDATE(bin_ver) },
-	{ "1st MAC Address",           6, DEFINE_PRINT_UPDATE(mac) },
-	{ "2nd MAC Address",           6, DEFINE_PRINT_UPDATE(mac) },
-	{ "Production Date",           4, DEFINE_PRINT_UPDATE(date) },
-	{ "Serial Number",            12, DEFINE_PRINT_UPDATE(bin_rev) },
-	{ "3rd MAC Address (WIFI)",    6, DEFINE_PRINT_UPDATE(mac) },
-	{ "4th MAC Address (Bluetooth)", 6, DEFINE_PRINT_UPDATE(mac) },
-	{ "Layout Version",            1, DEFINE_PRINT_UPDATE(bin) },
-	{ RESERVED_FIELDS,            83, DEFINE_PRINT_UPDATE(reserved) },
-	{ "Product Name",             16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #1",       16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #2",       16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #3",       16, DEFINE_PRINT_UPDATE(ascii) },
-	{ RESERVED_FIELDS,            64, eeprom_field_print_reserved,
-					  eeprom_field_update_ascii },
+	{ "Major Revision",            2, DEFINE_FIELD_FUNC(bin_ver) },
+	{ "Minor Revision",            2, DEFINE_FIELD_FUNC(bin_ver) },
+	{ "1st MAC Address",           6, DEFINE_FIELD_FUNC(mac) },
+	{ "2nd MAC Address",           6, DEFINE_FIELD_FUNC(mac) },
+	{ "Production Date",           4, DEFINE_FIELD_FUNC(date) },
+	{ "Serial Number",            12, DEFINE_FIELD_FUNC(bin_rev) },
+	{ "3rd MAC Address (WIFI)",    6, DEFINE_FIELD_FUNC(mac) },
+	{ "4th MAC Address (Bluetooth)", 6, DEFINE_FIELD_FUNC(mac) },
+	{ "Layout Version",            1, DEFINE_FIELD_FUNC(bin) },
+	{ RESERVED_FIELDS,            83, DEFINE_FIELD_FUNC(reserved) },
+	{ "Product Name",             16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #1",       16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #2",       16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #3",       16, DEFINE_FIELD_FUNC(ascii) },
+	{ RESERVED_FIELDS,            64, FIELD_FUNC_RES_LAST },
 };
 
 struct eeprom_field layout_v3[16] = {
-	{ "Major Revision",            2, DEFINE_PRINT_UPDATE(bin_ver) },
-	{ "Minor Revision",            2, DEFINE_PRINT_UPDATE(bin_ver) },
-	{ "1st MAC Address",           6, DEFINE_PRINT_UPDATE(mac) },
-	{ "2nd MAC Address",           6, DEFINE_PRINT_UPDATE(mac) },
-	{ "Production Date",           4, DEFINE_PRINT_UPDATE(date) },
-	{ "Serial Number",            12, DEFINE_PRINT_UPDATE(bin_rev) },
-	{ "3rd MAC Address (WIFI)",    6, DEFINE_PRINT_UPDATE(mac) },
-	{ "4th MAC Address (Bluetooth)", 6, DEFINE_PRINT_UPDATE(mac) },
-	{ "Layout Version",            1, DEFINE_PRINT_UPDATE(bin) },
-	{ "CompuLab EEPROM ID",        3, DEFINE_PRINT_UPDATE(bin) },
-	{ RESERVED_FIELDS,            80, DEFINE_PRINT_UPDATE(reserved) },
-	{ "Product Name",             16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #1",       16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #2",       16, DEFINE_PRINT_UPDATE(ascii) },
-	{ "Product Options #3",       16, DEFINE_PRINT_UPDATE(ascii) },
-	{ RESERVED_FIELDS,            64, eeprom_field_print_reserved,
-					  eeprom_field_update_ascii },
+	{ "Major Revision",            2, DEFINE_FIELD_FUNC(bin_ver) },
+	{ "Minor Revision",            2, DEFINE_FIELD_FUNC(bin_ver) },
+	{ "1st MAC Address",           6, DEFINE_FIELD_FUNC(mac) },
+	{ "2nd MAC Address",           6, DEFINE_FIELD_FUNC(mac) },
+	{ "Production Date",           4, DEFINE_FIELD_FUNC(date) },
+	{ "Serial Number",            12, DEFINE_FIELD_FUNC(bin_rev) },
+	{ "3rd MAC Address (WIFI)",    6, DEFINE_FIELD_FUNC(mac) },
+	{ "4th MAC Address (Bluetooth)", 6, DEFINE_FIELD_FUNC(mac) },
+	{ "Layout Version",            1, DEFINE_FIELD_FUNC(bin) },
+	{ "CompuLab EEPROM ID",        3, DEFINE_FIELD_FUNC(bin) },
+	{ RESERVED_FIELDS,            80, DEFINE_FIELD_FUNC(reserved) },
+	{ "Product Name",             16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #1",       16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #2",       16, DEFINE_FIELD_FUNC(ascii) },
+	{ "Product Options #3",       16, DEFINE_FIELD_FUNC(ascii) },
+	{ RESERVED_FIELDS,            64, FIELD_FUNC_RES_LAST },
 };
 
 void eeprom_layout_assign(struct eeprom_layout *layout, int layout_version)
diff --git a/common/eeprom/eeprom_field.c b/common/eeprom/eeprom_field.c
index 7099b8d8b4..42e1d7ddeb 100644
--- a/common/eeprom/eeprom_field.c
+++ b/common/eeprom/eeprom_field.c
@@ -130,6 +130,28 @@ int eeprom_field_update_bin(struct eeprom_field *field, uchar *fbuf,
 	return __eeprom_field_update_bin(field, fbuf, value, false);
 }
 
+/**
+ * eeprom_field_read_bin() - Read field data in binary form
+ *
+ * @field:     an initialized field
+ * @fbuf:      field buffer
+ * @buf:       read buffer
+ * @buf_size:  read buffer size
+ */
+int eeprom_field_read_bin(const struct eeprom_field *field, uchar *fbuf,
+                         uchar *buf, int buf_size)
+{
+	memset(buf, 0, buf_size);
+	if (field->size <= 0)
+		return 0;
+	if (field->size > buf_size)
+		memcpy(buf, fbuf, buf_size);
+	else
+		memcpy(buf, fbuf, field->size);
+
+	return 0;
+}
+
 /**
  * eeprom_field_update_reserved() - Update reserved field with new data in
  *				    binary form
diff --git a/common/eeprom/eeprom_layout.c b/common/eeprom/eeprom_layout.c
index 363d8bb96d..3d8b4ec1b6 100644
--- a/common/eeprom/eeprom_layout.c
+++ b/common/eeprom/eeprom_layout.c
@@ -96,6 +96,42 @@ static int eeprom_layout_update_field(struct eeprom_layout *layout,
 	return -1;
 }
 
+/*
+ * eeprom_layout_read_field() - read a single field from the layout data
+ * @layout:    A pointer to an existing struct layout.
+ * @field_name:        The name of the field to update.
+ * @buf:       Read data buffer.
+ * @buf_size:  Read data buffer size.
+ *
+ * Returns: 0 on success, negative error value on failure.
+ */
+static int eeprom_layout_read_field(struct eeprom_layout *layout,
+				    char *field_name, uchar *buf, int buf_size)
+{
+       int i, err;
+       struct eeprom_field *fields = layout->fields;
+       uchar *fbuf = layout->data;
+
+       if ((field_name == NULL) || (buf == NULL) || (buf_size <= 0))
+               return -1;
+
+	for (i = 0; i < layout->num_of_fields; fbuf += fields[i++].size) {
+		if (fields[i].name == RESERVED_FIELDS ||
+		    strcmp(fields[i].name, field_name))
+			continue;
+
+		err = fields[i].read(&fields[i], fbuf, buf, buf_size);
+		if (err)
+			printf("Field %s, read error\n", field_name);
+
+		return err;
+	}
+
+	printf("No such field '%s'\n", field_name);
+
+	return -1;
+}
+
 /*
  * eeprom_layout_setup() - setup layout struct with the layout data and
  *			   metadata as dictated by layout_version
@@ -117,4 +153,5 @@ void eeprom_layout_setup(struct eeprom_layout *layout, unsigned char *buf,
 	layout->data_size = buf_size;
 	layout->print = eeprom_layout_print;
 	layout->update = eeprom_layout_update_field;
+	layout->read = eeprom_layout_read_field;
 }
diff --git a/include/eeprom_field.h b/include/eeprom_field.h
index 0ada7f3cfe..d068980fdd 100644
--- a/include/eeprom_field.h
+++ b/include/eeprom_field.h
@@ -19,11 +19,15 @@ struct eeprom_field {
 	void (*print)(const struct eeprom_field *eeprom_field, uchar *fbuf);
 	int (*update)(struct eeprom_field *eeprom_field, uchar *fbuf,
 		      char *value);
+	int (*read)(const struct eeprom_field *field, uchar *fbuf, uchar *buf,
+		    int buf_size);
 };
 
 void eeprom_field_print_bin(const struct eeprom_field *field, uchar *fbuf);
 int eeprom_field_update_bin(struct eeprom_field *field, uchar *fbuf,
 			    char *value);
+int eeprom_field_read_bin(const struct eeprom_field *field, uchar *fbuf,
+			  uchar *buf, int buf_size);
 
 void eeprom_field_print_bin_rev(const struct eeprom_field *field, uchar *fbuf);
 int eeprom_field_update_bin_rev(struct eeprom_field *field, uchar *fbuf,
diff --git a/include/eeprom_layout.h b/include/eeprom_layout.h
index 459b99d861..808a458ab0 100644
--- a/include/eeprom_layout.h
+++ b/include/eeprom_layout.h
@@ -23,6 +23,8 @@ struct eeprom_layout {
 	void (*print)(const struct eeprom_layout *eeprom_layout);
 	int (*update)(struct eeprom_layout *eeprom_layout, char *field_name,
 		      char *new_data);
+	int (*read)(struct eeprom_layout *eeprom_layout, char *field_name,
+		    uchar *buf, int buf_size);
 };
 
 void eeprom_layout_setup(struct eeprom_layout *layout, unsigned char *buf,
-- 
2.17.1

